platform/cxxsupport/mstd_span

```diff
@@ -0,0 +1,381 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MSTD_SPAN_
+#define MSTD_SPAN_
+
+#if __cplusplus >= 201703L && __has_include(<span>)
+#include <version>
+#endif
+
+#if __cpp_lib_span >= 202002L
+#include <span>
+
+namespace mstd {
+    using std::span;
+}
+
+#else //__cpp_lib_span >= 202002L
+
+#include <array>
+#include <mstd_iterator>
+#include <mstd_type_traits>
+#include "mbed_assert.h"
+
+namespace mstd {
+
+constexpr size_t dynamic_extent = -1;
+
+template <typename, size_t = dynamic_extent> class span;
+
+namespace detail {
+
+template<typename ElementType, size_t Extent>
+class storage {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t) noexcept :
+        _data(ptr)
+    {}
+
+    ElementType *_data = nullptr;
+    static constexpr size_t _size = Extent;
```

Comment:
You need a definition of this outside the class (for pre-C++17).

(Reference: https://en.cppreference.com/w/cpp/language/static - section "Constant static members")

Personally, I'd make the storage have a `size()` method, rather one having a "fake" `_size`, but I guess this is fine.

Reply:
Seems to me that it's valid as it is? It says if a constexpr data member is odr used we need a definition at namespace scope. But since we only read the constant variable we don't have an odr use and should be fine with this.

Reply:
Looking at this again, there is also this:

> (since c++11)
> If a static data member of LiteralType is declared constexpr, it must be initialized with an initializer in which every expression is a constant expression, right inside the class definition: 

Since extent is always known at compile time we should be fine here.

Reply:
See section "ODR use" here: https://en.cppreference.com/w/cpp/language/definition%23One_Definition_Rule

Reading the variable is a use. `sizeof` it is not.

This isn't a totally theoretically piece of nitpicking. If compiled with `-O0` current compilers can actually try to look at the value at run-time, despite the fact they "know" it at compiler time. They're not required to do that optimisation. We've had these lack-of-definition errors pop up occasionally for people doing debug builds.

Now, if you're just directly using the value in a simple way it looks like GCC and Clang will just use the constant even at O0, so you'd get away with it.

But here's an example of a compiler needing the definition, triggered by reference parameters:

https://godbolt.org/z/Kfa16z

That will not link, as `my_constant` is not defined so it can't pass its address to `std::min`. Change that to -O2 and you'd be okay.

(Same results on both Clang and GCC.)

Reply:
I was genuinely trying to understand why this would fail because I couldn't get it to fail but your explanation is clear.

If I open your example it does build with the default settings but I guess you mean it only names the variable without initializing it.

So this is an actual working example then:
https://godbolt.org/z/vbaxf8

I'll add something similar for both occurrences

Reply:
Godbolt only does compilation, not linking. As you can see in the assembly, it is passing the address of the constant to `std::min`  (`_ZN7MyClassILi23EE11my_constantE`, loaded into R0).  That would fail if you tried to link it, due to that not being defined anywhere.

Yes, your example shows that you're providing the definition as well as referencing it, so it will link.

The actual definition would appear in every object file using `MyClass<23>`, but it would be a common definition - they'd all unite to be "one definition", and the linker would drop it from the final binary if never actually used, due to compiler optimisation.